contracts/root/withdrawManager/WithdrawManager.sol

```diff
@@ -47,142 +52,229 @@ contract WithdrawManager is WithdrawManagerStorage, IWithdrawManager {
+
+    WithdrawManagerHeader.PlasmaExit memory _exitObject = PlasmaExit({
+      owner: msg.sender,
+      token: _token,
+      receiptAmountOrNFTId: _amountOrNFTId,
+      burnt: false
+    });
+
+    _addExitToQueue(
+      _exitObject,
+      _depositBlockId * HEADER_BLOCK_NUMBER_WEIGHT, // exit id
+      _createdAt
     );
   }
 
-  function _processBurntTx(
-    bytes memory txBytes, bytes memory path, bytes memory txProof, bytes32 txRoot,
-    address rootToken, uint256 amountOrTokenId, address sender)
+  function _withdraw(
+    PlasmaExit memory exitObject,
+    uint256 headerNumber,
+    bytes memory withdrawBlockProof,
+    uint256 withdrawBlockNumber,
+    uint256 withdrawBlockTime,
+    bytes32 withdrawBlockTxRoot,
+    bytes32 withdrawBlockReceiptRoot,
+    bytes memory path,
+    uint8 oIndex)
     internal
-    view
   {
-    // check basic tx format
-    RLPReader.RLPItem[] memory txList = txBytes.toRlpItem().toList();
-    require(txList.length == 9, "MALFORMED_WITHDRAW_TX");
+    uint256 startBlock;
+    bytes32 headerRoot;
+
+    // @todo a function to return just root and startBlock might save gas
+    (headerRoot, startBlock,,,) = rootChain.headerBlocks(headerNumber);
 
-    // check mapped root<->child token
     require(
-      registry.rootToChildToken(rootToken) == txList[3].toAddress(),
-      "INVALID_ROOT_TO_CHILD_TOKEN_MAPPING"
+      keccak256(abi.encodePacked(withdrawBlockNumber, withdrawBlockTime, withdrawBlockTxRoot, withdrawBlockReceiptRoot))
+        .checkMembership(withdrawBlockNumber - startBlock, headerRoot, withdrawBlockProof),
+      "WITHDRAW_BLOCK_NOT_A_PART_OF_SUBMITTED_HEADER"
     );
 
-    require(txList[5].toBytes().length == 36, "MALFORMED_WITHDRAW_TX");
-    // check withdraw function signature
-    require(
-      BytesLib.toBytes4(BytesLib.slice(txList[5].toBytes(), 0, 4)) == WITHDRAW_SIGNATURE,
-      "WITHDRAW_SIGNATURE_NOT_FOUND"
+    uint256 _exitId = (
+      headerNumber * HEADER_BLOCK_NUMBER_WEIGHT +
+      withdrawBlockNumber * WITHDRAW_BLOCK_NUMBER_WEIGHT +
+      path.toRlpItem().toBytes().toRlpItem().toUint() * 100000 +
+      oIndex
     );
-    require(registry.isERC721(rootToken) || amountOrTokenId > 0);
-    require(amountOrTokenId == BytesLib.toUint(txList[5].toBytes(), 4));
 
-    // Make sure this tx is the value on the path via a MerklePatricia proof
+    _addExitToQueue(exitObject, _exitId, withdrawBlockTime);
+  }
+
+  /**
+  * @dev Adds an exit to the exit queue.
+  * @param _exitObject Exit plasma object
+  * @param _exitId Position of the UTXO in the child chain (withdrawBlockNumber, txIndex, oIndex)
+  * @param _createdAt Time when the UTXO was created.
+  */
+  function _addExitToQueue(
+    PlasmaExit memory _exitObject,
+    uint256 _exitId,
+    uint256 _createdAt
+  ) internal {
     require(
-      MerklePatriciaProof.verify(txBytes, path, txProof, txRoot),
-      "INVALID_TX_MERKLE_PROOF"
+      exits[_exitId].token == address(0x0),
+      "EXIT_ALREADY_EXISTS"
```

Comment:
this check isn't about EXIT, it checks that token is mapped and there exists a PQ for it, so maybe something in that context we need a require message.